### PR TITLE
Delete compact block on reload error

### PR DESCRIPTION
**Pull request by krasi-georgiev**
To avoid a dead loop a failed reload immediately after a compaction should delete the
resulting block to avoid creating blocks with the same time range.

fixes https://github.com/prometheus/prometheus/issues/5135

Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

**Comment by brian-brazil** [comment]
Wouldn't this leave us with a race if the Prometheus happened to die before the deletion happened?

**Comment by codesome** [comment]
Continuing what @brian-brazil said, how about having a boolean field in the meta (`loaded`?) - `false` would mean "it was compacted and not loaded even once".

Would this make it simple to handle the race?

**Comment by brian-brazil** [comment]
That'd mean that we'd be changing blocks after they're written, which I don't think we should start doing.

**Comment by krasi-georgiev** [comment]
I don't see a race here.
Prometheus will die and will load the block as normal at the next start up.

**Comment by krasi-georgiev** [comment]
most changes that look unrelated are for the windows tests to pass.

**Comment by codesome** [comment]
> I don't see a race here.

On a second look now, feels the same.

**Comment by krasi-georgiev** [comment]
@brian-brazil Do you see a race somewhere else? 
re my comment at  https://github.com/prometheus/tsdb/pull/512#issuecomment-458878260

File: compact_test.go [hunk 1]
```diff
@@ -741,3 +745,78 @@ func TestDisableAutoCompactions(t *testing.T) {
 	}
 	testutil.Assert(t, len(db.Blocks()) > 0, "No block was persisted after the set timeout.")
 }
+
+// TestDeleteCompactionBlockAfterFailedReload ensures that a failed reload imidiately after a compaction
```
Comment (codesome) [inline compact_test.go]:
```suggestion
// TestDeleteCompactionBlockAfterFailedReload ensures that a failed reload immediately after a compaction
```

File: db_test.go [hunk 2]
```diff
@@ -429,11 +448,11 @@ func TestDB_Snapshot(t *testing.T) {
 	// reopen DB from snapshot
 	db, err = Open(snap, nil, nil, nil)
 	testutil.Ok(t, err)
-	defer db.Close()
+	defer func() { testutil.Ok(t, db.Close()) }()
 
 	querier, err := db.Querier(mint, mint+1000)
 	testutil.Ok(t, err)
-	defer querier.Close()
+	defer func() { querier.Close() }()
```
Comment (codesome) [inline db_test.go]: Did you mean to have `testutil.Ok(t, querier.Close())`?

**Comment by communityexplorer** [comment]
how can I know which file should I delete?

**Comment by codesome** [comment]
@communityexplorer Can you give us some context?

File: compact_test.go [hunk 3]
```diff
@@ -741,3 +745,78 @@ func TestDisableAutoCompactions(t *testing.T) {
...
+			testutil.Equals(t, expBlocks, len(actBlocks)-1) // -1 to exclude the corrupted block.
+			testutil.Ok(t, os.RemoveAll(lastBlockIndex))    // Corrupt the block by removing the index file.
+
+			testutil.Equals(t, 0.0, prom_testutil.ToFloat64(db.metrics.reloadsFailed), "initial 'failed db reload' count metrics mismatch")
+			testutil.Equals(t, 0.0, prom_testutil.ToFloat64(db.compactor.(*LeveledCompactor).metrics.ran), "initial `compactions` count metric mismatch")
+
+			// Do the compaction and check the metrics.
+			// Compaction should succeed, but the reload should fail and
+			// the new block created from the compaction should be deleted.
+			db.EnableCompactions()
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply compact_test.go]: thanks 
updated!

File: compact_test.go [hunk 4]
```diff
@@ -741,3 +745,78 @@ func TestDisableAutoCompactions(t *testing.T) {
...
+			})
+			defer func() {
+				testutil.Ok(t, db.Close())
+				delete()
+			}()
+			db.DisableCompactions()
+
+			expBlocks := bootStrap(db)
+
+			// Create a block that will trigger the reloard to fail.
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply compact_test.go]: thanks 
updated!

File: compact_test.go (hunk 1 above)
Comment (krasi-georgiev) [reply compact_test.go]: thanks 
updated!

File: db_test.go (hunk 2 above)
Comment (krasi-georgiev) [reply db_test.go]: thanks 
updated!

**Comment by krasi-georgiev** [comment]
> We need to do it in `db.CleanTombstones()` too. It creates new blocks and calls `db.reload()`.

good thought. I just double checked and this is already handled.

https://github.com/prometheus/tsdb/blob/0b72f9af4c7a7953f89c993e69bbf81b12af49c3/db.go#L964-L971

File: block.go [hunk 5]
```diff
@@ -269,10 +270,18 @@ type Block struct {
...
 // to instantiate chunk structs.
-func OpenBlock(logger log.Logger, dir string, pool chunkenc.Pool) (*Block, error) {
+func OpenBlock(logger log.Logger, dir string, pool chunkenc.Pool) (pb *Block, err error) {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
+	var closers []io.Closer
+	defer func() {
+		if err != nil {
+			for _, c := range closers {
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply block.go]: thanks. Updated!

**Comment by codesome** [comment]
Apparently, it was my PR https://github.com/prometheus/tsdb/pull/341 :)

**Comment by communityexplorer** [comment]
> @communityexplorer Can you give us some context?
as we know, it recreated many repeated blocks, how can we know  which blocks should we delete, is it necessary to restart prometheus after that, how can we avoid this promblem?

**Comment by codesome** [comment]
@communityexplorer In this PR the blocks to be deleted are known (the one being created in compaction), and it is done automatically.

**Comment by communityexplorer** [comment]
@codesome @krasi-georgiev 
when will this solved version release?  my PR is #5120, this problem has been frequent recently in our production env, can you give release as soon as possible?

**Comment by communityexplorer** [comment]
another question:
(memory size) = vm.max_map_count * (memory chunk)
what is the memory chunk size for prometheus?
thanks so much

**Comment by simonpasquier** [comment]
@communityexplorer this fix should be available in Prometheus v2.8.

**Pull request merged**